iot-hub/Samples/device/DeviceReconnectionSample/DeviceReconnectionSample.cs

```diff
@@ -98,101 +116,96 @@ private void ConnectionStatusChangeHandler(ConnectionStatus status, ConnectionSt
-                eventArgs.Cancel = true;
-                cts.Cancel();
-                _logger.LogInformation("Sample execution cancellation requested, will exit.");
-            };
-
-            try
-            {
-                await Task.WhenAll(SendMessagesAsync(cts.Token), ReceiveMessagesAsync(cts.Token));
-            }
-            catch (Exception ex)
-            {
-                _logger.LogError($"Unrecoverable exception caught, user action is required, so exiting...: \n{ex}");
-            }
-        }
-
-        private void InitializeClient()
+        private async Task InitializeClient()
         {
-            // If the client reports Connected status, it is already in operational state.
-            if (s_connectionStatus != ConnectionStatus.Connected)
+            if (ClientShouldBeInitialized(s_connectionStatus))
             {
-                lock (_lock)
+                // Allow a single thread to dispose and initialize the client instance.
+                await s_initializeClientSemaphore.WaitAsync();
+                if (ClientShouldBeInitialized(s_connectionStatus))
                 {
                     _logger.LogDebug($"Attempting to initialize the client instance, current status={s_connectionStatus}");
 
-                    // If the device client instance has been previously initialized, then dispose it.
-                    // The s_wasEverConnected variable is required to store if the client ever reported Connected status.
-                    if (s_wasEverConnected && s_connectionStatus == ConnectionStatus.Disconnected)
+                    if (s_connectionStatus == ConnectionStatus.Disconnected)
                     {
-                        s_deviceClient?.Dispose();
-                        s_wasEverConnected = false;
+                        // If the device client instance has been previously initialized, then dispose it.
+                        if (s_deviceClient != null)
+                        {
+                            s_deviceClient.Dispose();
+                            s_deviceClient = null;
+                        }
                     }
 
                     s_deviceClient = DeviceClient.CreateFromConnectionString(_deviceConnectionString, _transportType);
                     s_deviceClient.SetConnectionStatusChangesHandler(ConnectionStatusChangeHandler);
-                    s_deviceClient.OperationTimeoutInMilliseconds = (uint)s_operationTimeout.TotalMilliseconds;
+
                     _logger.LogDebug($"Initialized the client instance.");
                 }
+                s_initializeClientSemaphore.Release();
             }
+
+            // OpenAsync() is an idempotent call, it has the same effect if called once or multiple times on the same client.
+            await s_deviceClient.OpenAsync();
+            _logger.LogDebug($"Successfully opened the client instance.");
         }
 
         private async Task SendMessagesAsync(CancellationToken cancellationToken)
```

Comment:
I'd suggest separating the code to send telemetry from this client lifetime manager. It could offer a method to send telemetry by other parts of the application, but it seems naïve to think all the code will live in one place.

That matters because this is a sample, and we should try to follow what a custom would do.

Reply:
yes, that is the approach I am trying out now; encapsulate all lifecycle modifying code into an `IotHubClient`, which holds a reference to a device client. The sample will then call `IotHubClient.sendEvent()` etc, which will delegate the actual operation to the device client.